### ISSUE TITLE
FontForge scripting cannot change glyphs' width precisely

### ISSUE THREAD

**Issue by aerobounce** [fontforge/fontforge#4035]
First of all, excuse me if this is not a FontForge specific problem.

----

- [x] Screenshot
- [x] The FontForge version and the operating system you're using
- [x] The behavior you expect to see, and the actual behavior
- [x] Steps to reproduce the behavior 
- [ ] \(optional) Possible solution/fix/workaround

----

## Env
OS: macOS 10.14.6 (18G1012)
fontforge: Version: 20190801, Based on sources from 12:52 UTC 14-Nov-2019-D.
_Using 'ff' Script mode_

## Problem
Simple question.
I'm trying to tighten fonts' letter spacing. It works, but not precisely.
The fonts I'm modifying is `SF Mono` which you can download freely: https://developer.apple.com/fonts/

## Steps to reproduce
Okay, now please have a look at those screenshots:

![before](https://user-images.githubusercontent.com/10491362/69656219-81c01780-10bb-11ea-82aa-b668c5d431ab.jpg)
![after](https://user-images.githubusercontent.com/10491362/69656220-81c01780-10bb-11ea-9426-6c77e7a9550b.jpg)

You can see, at `1215`, glyphs gets transformed. 
**1215** is some kind of boundary line, if it's `1216`, nothing happens.
I want something like `1230` to work.

I also tried adjusting Left / Right Bearings, but it was the same. There's some boundaries that fonts get react and I cannot get precise numbers to work.

## Expected behavior
In this case, `1215` is a bit too tight for me.  
I'd like to see something like glyph's width = `1230` to work, but I can never achieve that.

## Question
Is this a Fontforge bug, or is that how Fonts works?
One thing I am supposing is how font is created and depending on that some kind of boundaries exist (depending on EM, Ascent, Descent... etc).

For SF Mono font, that's the boundary line that you cannot avoid?

Thanks.

**Comment by aerobounce** [comment]
I did the same thing with GUI FontForge and result was the same.

**Comment by skef** [comment]
I just ran your script on SF-Mono-Regular.otf to produce a new .otf file with modified widths and it produced a font with the expected modified width.

There are a number of problems you might be seeing. The tool you are using to test might impose its own spacing. There could be cached copies of the modified font in your system keyed to its name or some other identifying token. One way to distinguish between the two is to modify a character in each of your test fonts to see if it's reproduced as you expect. 

Another thing to try is to test with a very large point size, which will usually make a tool less likely to impose its own heuristics and rendering tricks.

**Comment by aerobounce** [comment]
> I just ran your script on SF-Mono-Regular.otf to produce a new .otf file with modified widths and it produced a font with the expected modified width.

That's very interesting, with any widths you specify? Such as very one I want, around `1230`?

> There are a number of problems you might be seeing. The tool you are using to test might impose its own spacing. There could be cached copies of the modified font...

This is also an interesting point and never thought about that.
Yes, I'm using Sublime Text 3 to check the result and ST3 has some kind of unique rendering actually. Well this will be worth trying for sure.

**Comment by skef** [comment]
I tried various values but just tried 1230 and it worked just as well. Note that I'm not "testing", I'm just generating an `.otf` and verifying the metrics of a few characters. There are no kerning tables (as usual for a monospace font) so those can't be getting in the way.

**Comment by aerobounce** [comment]
Ah, you mean you checked literal values of generated fonts, not 'tested' rendering of it. I see.

**Comment by skef** [comment]
Let's put it this way: All FontForge can do is manipulate the spacing parameters, and it looks like that's happening in this case, in my testing no differently for the values that aren't working for you than for the values that are.

You might try uploading the fonts you generate to something like http://webfont-test.com/ , which should give you a "clean", non-cached result every time. That way you could confirm that the file itself has the right values and then investigate why your testing yields a different result.

**Comment by aerobounce** [comment]
First of all thanks for the website, I didn't know such testing web-app exists.

Well, and thanks again for your all advices and quick responses. Those all were I couldn't come up with and I didn't know.
And as you mentioned, I tried rendering generated fonts with other app, like Terminal. And the result was quite different.

It's been only just an hour, but **I've made quite satisfactionary improvements!** And getting much closer to what I'd like to achieve. Kind of making more sense to me now, how `width` works.

I think I can close the issue now, as I can do more investigating based on your advices, and more importantly it was not FontForge bug at all.
Okay, thanks again. Let me note you when the font I'm working on is released. :)
